### ISSUE TITLE
Appsignal logging "Nothing on timestack in finish event"

### ISSUE THREAD

**Issue by alanhala** [appsignal/appsignal-ruby#672]
I was investigating the logs of the app I'm working on and I found that at the time we had an outage, appsignal was logging lots of entries like this one:
`[2020-12-28T23:35:36 (extension) #3][Error] appsignal: Nothing on timestack in finish event`

We are using
- appsignal (2.11.1)
- ruby 2.6.6
- rails 6.0.3.4

I tried to investigate and find where this entry might have been logged but I couldn't. So I'm creating this issue to understand the cause of that log entry.

Thanks!

**Comment by alanhala** [comment]
anything?

**Comment by tombruijn** [comment]
Hi @alanhala, this issue got missed by us. I've added it to our board to look at.

From what I can tell so far, This error comes from our extension when an event is "finished" here: https://github.com/appsignal/appsignal-ruby/blob/889ae6d855a4bbe15ff498d386293c75885c1eaa/lib/appsignal/transaction.rb#L349
but the event was never registered as "started".

Is there anywhere in the app that the AppSignal `finish_event` or `instrument` methods are called?

**Comment by alanhala** [comment]
Hi @tombruijn

No, we don't explicitly call any of those methods in our app if that's what you meant.

**Comment by tombruijn** [comment]
That is what I was wondering. Ok then. It shouldn't break anything. It's just that somewhere an event doesn't get "finished" properly. I don't see how this could be the case in the instrumentation we provide with the rails middleware, so I assumed it would be coming from some custom instrumentation (https://docs.appsignal.com/ruby/instrumentation/instrumentation.html).

**Comment by alanhala** [comment]
Is there a way to know which event is the one that is breaking?

**Comment by tombruijn** [comment]
Not currently no. That should be part of the error message, which I may change right now, but can only be seen with a new extension build and Ruby gem release. I had a look if our debug log mode logs more data, but it only does that if it gets past this error. The only other way is to isolate the call in the app, but that would mean you already know what is causing it.

**Comment by alanhala** [comment]
Great! So you will release a new gem version adding more information to the message?

**Comment by tombruijn** [comment]
Yes, but I'm not sure when yet. It's annoying the error message is logged, but it doesn't look like something is horribly broken unless the log is filled with them for every minute. And then only the events aren't recorded properly.

**Comment by tombruijn** [comment]
@alanhala I've released 2.11.5 with more logging. Can you update the gem in your app and see what extra information is being logged on that same error line?

**Comment by alanhala** [comment]
@tombruijn Great! Will do. Thanks.

**Comment by tombruijn** [comment]
@alanhala can you also enable [transaction_debug_mode](https://docs.appsignal.com/ruby/configuration/options.html#option-transaction_debug_mode)? That way we have even more information to debug it.

**Comment by alanhala** [comment]
@tombruijn I've just updated the gem and waiting to see if the log shows up again. By enabling `transaction_debug_mode` we will start logging a lot more, right? Not sure I will be able to do that in production. I would prefer waiting and see if your new release adds the information I need to figure out where's the problem.

**Comment by tombruijn** [comment]
@alanhala Yes it will log a bit more information yes. It logs every call being made to our extension, including the start_event and finish_event functions. The more requests/jobs/etc the more transactions. But I think we'll need that extra information to see which event didn't get finished properly.

**Comment by alanhala** [comment]
@tombruijn we've just received the extended log entry. Here's an example:
`[2021-02-03T21:23:47 (extension) #3][Error] appsignal: Nothing on timestack in finish event for event 69db753f-8db0-487f-94a0-5a42e91fcad5 - perform_job.atlas`

`perform_job.atlas` is what we use when we execute jobs in our background system, similar to sidekiq. This is what we do:
```ruby
Appsignal.monitor_transaction(
  "perform_job.atlas",
  class: job.class,
  method: "perform",
  queue: job.queue_name,
  queue_start: job.scheduled_at.to_i,
  params: job_data
) do
  # Do stuff
end
```

**Comment by alanhala** [comment]
> @alanhala Yes it will log a bit more information yes. It logs every call being made to our extension, including the start_event and finish_event functions. The more requests/jobs/etc the more transactions. But I think we'll need that extra information to see which event didn't get finished properly.

As you see, it took a few days to happen. As it only happens in production, I'm not able to add the debug mode as it will be a massive amount of logs :/

**Comment by tombruijn** [comment]
@alanhala thanks for the new log entry. It's not the `perform_job.atlas` event specifically that's the problem probably, but this helps us find the event that does. It is just the last event in the stack that is trying to close, but can't. More events have been finished than have been started.

The quickest way to debug this issue would be enabling the `transaction_debug_mode`. That way we can see all the extension function calls and see which event specifically is causing the issue. So please see if you can find a way to enable it for just long enough for this log entry to trigger while transaction debug mode is on.

As the app is using some custom instrumentation my guess is that something is happening there. We both capture `raise` and `throw` calls in both the `monitor_transaction` and `instrument` helpers so it shouldn't be an unhandled exception or signal that's the problem. You say the app doesn't use any other `Appsignal.instrument` method calls, correct? It also doesn't nest any `Appsignal.monitor_transaction` calls? Or is any [ActiveSupport::Notifications](https://api.rubyonrails.org/classes/ActiveSupport/Notifications.html) instrumentation (standard in Rails and some other gems) present?

Where is the `monitor_transaction` called? It's not part of a gem we [already instrument](https://docs.appsignal.com/ruby/integrations/)?

**Comment by alanhala** [comment]
@tombruijn hey sorry for the delay here.

I'm afraid I won't be able to enable the transaction_debug_mode in production :/.

> You say the app doesn't use any other Appsignal.instrument method calls, correct? It also doesn't nest any Appsignal.monitor_transaction calls? Or is any ActiveSupport::Notifications instrumentation (standard in Rails and some other gems) present?

No, we don't nest any monitor_transaction calls. We do have some ActiveSupport::Notifications instrumentation from Rails and from other gems like HTTP calls.

> Where is the monitor_transaction called? It's not called in a worker of a gem we already instrument?

We have our own background jobs system that it's built on top of ActiveJob

**Comment by tombruijn** [comment]
> We have our own background jobs system that it's built on top of ActiveJob

@alanhala  That is probably the cause of the log. In Ruby gem 2.11 we [have improved Active Job support](https://blog.appsignal.com/2020/08/11/ruby-gem-2-11.html) in such a way that it should auto instrument all adapters, not just the ones we have implemented Active Job support for (like Sidekiq, Delayed Job, etc).

It automatically creates a transaction for every Active Job job, so starting one manually with `Appsignal.monitor_transaction` isn't necessary and will create a new transaction, replacing the one that was already started.

To fix this, replace the `Appsignal.monitor_transaction` with an `Appsignal.instrument` to only register an event (or leave it out entirely) it should then not have the issue anymore hopefully.

**Comment by alanhala** [comment]
@tombruijn we are not using `ApplicationJob.execute` for performing the jobs, so we are not using the Appsignal hook. That means that we need to call `Appsignal.monitor_transaction`. Also I checked that `Appsignal.monitor_transaction` calls `Appsignal::Transaction.create` which doesn't replace the transaction if there's a current one. So maybe that's not the issue here?

**Comment by tombruijn** [comment]
@alanhala  `Appsignal.monitor_transaction` doesn't replace the current transaction, but it does reuse it. Do you see a debug log message starting with "Trying to start new transaction with id" in the appsignal.log? Can you send your appsignal.log or a full [diagnose report](https://docs.appsignal.com/ruby/command-line/diagnose.html) to support@appsignal.com ?

When it reuses a trnasaction, when the `Appsignal.monitor_transaction` block ends it "completes" the transaction and no more calls can be made on it. My thought is that when the last event gets closed after `Appsignal.monitor_transaction` it can no longer be found because the parent transaction is already completed.

I hope the logs tell me more otherwise I'll need some kind of reproducible set up to figure out what's going on.

**Comment by alanhala** [comment]
> @alanhala `Appsignal.monitor_transaction` doesn't replace the current transaction, but it does reuse it. Do you see a warning log message starting with "Trying to start new transaction with id" in the appsignal.log? Can you send your appsignal.log or a full [diagnose report](https://docs.appsignal.com/ruby/command-line/diagnose.html) to [support@appsignal.com](mailto:support@appsignal.com) ?
> 
> When it reuses a transaction, when the `Appsignal.monitor_transaction` block ends it "completes" the transaction and no more calls can be made on it. My thought is that when the last event gets closed after `Appsignal.monitor_transaction` it can no longer be found because the parent transaction is already completed.
> 
> I hope the logs tell me more otherwise I'll need some kind of reproducible set up to figure out what's going on.

No, no warnings with that message. I've already run the diagnose and sent it to support.

**Comment by tombruijn** [comment]
[Private support conversation](https://app.intercom.com/a/apps/yzor8gyw/inbox/inbox/mentions/conversations/16410700029624)

**Comment by tombruijn** [comment]
Published gem 3.0.3 with more debug logging.
